src/dotnet/commands/dotnet-add-p2p/Program.cs

```diff
@@ -0,0 +1,232 @@
+            }
+            catch (GracefulException e)
+            {
+                Reporter.Error.WriteLine(e.Message);
+                return 1;
+            }
+        }
+
+        // There is ProjectRootElement.TryOpen but it does not work as expected
+        // I.e. it returns null for some valid projects
+        public static ProjectRootElement TryOpenProject(string filename)
+        {
+            try
+            {
+                return ProjectRootElement.Open(filename);
+            }
+            catch (Microsoft.Build.Exceptions.InvalidProjectFileException)
+            {
+                return null;
+            }
+        }
+
+        public static ProjectRootElement GetProjectFromFileOrThrow(string filename)
+        {
+            if (!File.Exists(filename))
+            {
+                throw new GracefulException($"Provided project `{filename}` does not exist.");
+            }
+
+            var project = TryOpenProject(filename);
+            if (project == null)
+            {
+                throw new GracefulException($"Invalid MSBuild project `{filename}`.");
+            }
+
+            return project;
+        }
+
+        public static ProjectRootElement GetProjectFromCurrentDirectoryOrThrow()
+        {
+            DirectoryInfo currDir = new DirectoryInfo(Directory.GetCurrentDirectory());
+            FileInfo[] files = currDir.GetFiles("*proj");
+            if (files.Length == 0)
+            {
+                throw new GracefulException("Could not find any MSBuild project in the current directory.");
+            }
+
+            if (files.Length > 1)
+            {
+                throw new GracefulException("Found more than one MSBuild project in the current directory. Please specify which one to use.");
+            }
+
+            FileInfo projectFile = files.First();
+
+            if (!projectFile.Exists)
+            {
+                throw new GracefulException("Could not find any project in the current directory.");
+            }
+
+            var ret = TryOpenProject(projectFile.FullName);
+            if (ret == null)
+            {
+                throw new GracefulException($"Found an MSBuild project `{projectFile.FullName}` in the current directory but it is invalid.");
+            }
+
+            return ret;
+        }
+
+        public static Func<T, bool> AndPred<T>(params Func<T, bool>[] preds)
+        {
+            return (el) => preds.All((pred) => pred == null || pred(el));
+        }
+
+        public static string GetFrameworkConditionString(string framework)
+        {
+            if (string.IsNullOrEmpty(framework))
+            {
+                return null;
+            }
+
+            return $" '$(TargetFramework)' == '{framework}' ";
+        }
+
+        public static Func<T, bool> FrameworkPred<T>(string framework) where T : ProjectElement
+        {
+            string conditionStr = GetFrameworkConditionString(framework);
```

Comment:
I changed that to TryGetFrameworkConditionString which should clarify also this logic. Does that sound reasonable?